src/v2/Components/Payment/PaymentModal.tsx

```diff
@@ -0,0 +1,252 @@
+
+const logger = createLogger("Components/Payment/PaymentModal.tsx")
+
+const onCreditCardAdded = (
+  me: PaymentSection_me,
+  store: RecordSourceSelectorProxy<any>,
+  data: PaymentModalCreateCreditCardMutation["response"]
+): void => {
+  const {
+    createCreditCard: { creditCardOrError },
+  } = data
+
+  // Explicitly update the relay store to be aware of the new credit card
+  if (creditCardOrError.creditCardEdge) {
+    const meStore = store.get(me.id)
+    const connection = ConnectionHandler.getConnection(
+      meStore,
+      "PaymentSection_creditCards"
+    )
+    const mutationPayload = store.getRootField("createCreditCard")
+    const creditCardOrErrorEdge = mutationPayload.getLinkedRecord(
+      "creditCardOrError"
+    )
+    const creditCardEdge = creditCardOrErrorEdge.getLinkedRecord(
+      "creditCardEdge"
+    )
+    ConnectionHandler.insertEdgeAfter(connection, creditCardEdge)
+  }
+}
+
+const validator = (values: any) => {
+  const validationResult = validateAddress(values)
+  const errorsTrimmed = removeEmptyKeys(validationResult.errors)
+  return errorsTrimmed
+}
+
+const mutation = graphql`
+  mutation PaymentModalCreateCreditCardMutation($input: CreditCardInput!) {
+    createCreditCard(input: $input) {
+      creditCardOrError {
+        ... on CreditCardMutationSuccess {
+          creditCardEdge {
+            node {
+              ...UserSettingsPaymentsCreditCard @relay(mask: false)
+            }
+          }
+        }
+        ... on CreditCardMutationFailure {
+          mutationError {
+            type
+            message
+            detail
+          }
+        }
+      }
+    }
+  }
+`
+export interface PaymentModalProps {
+  show: boolean
+  closeModal: () => void
+  onSuccess?: () => void
+  onError?: (message: string) => void
+  me: PaymentSection_me
+  relay: RelayProp
+}
+
+export const PaymentModal: React.FC<PaymentModalProps> = props => {
+  const { show, closeModal, relay } = props
+  const stripe = useStripe()
+  const elements = useElements()
+  const [createError, setCreateError] = useState(null)
+
+  const addCreditCard = async values => {
+    const billingAddress: CreateTokenCardData = {
+      name: values.name,
+      address_line1: values.addressLine1,
+      address_line2: values.addressLine2,
+      address_city: values.city,
+      address_state: values.region,
+      address_zip: values.postalCode,
+      address_country: values.country,
+    }
+
+    const cardElement = elements.getElement(CardElement)
+    const stripeResults = await stripe.createToken(cardElement, billingAddress)
+    console.log("after create", stripeResults)
+    const { token, error } = stripeResults
+    if (error) {
+      setCreateError(error.message)
+    } else {
+      // TODO: potentially dry the redundant code in PaymentForm?
+      // console.log("commit", commitMutation)
+      // commitMutation("aaa")
+      commitMutation<PaymentModalCreateCreditCardMutation>(relay.environment, {
```

Comment:
We generally follow the pattern around force where mutations are placed in their own files/ mutations folder

Reply:
Yeah I was going to suggest this as well. It might make thinks easier to test also.

Reply:
Yes, I had it in a separate file initially. But this mutation needs almost all the component props plus those stripe stuff so ended up with a complex function with many parameters and thought maybe just dropping it back inside the component makes everything simpler. So either way, it didn't look that clean to me 😅

Reply:
Cool, either way 👍